공공데이터포털(www.data.go.kr) - 서울특별시 대기질 자료(미세먼지, 초미세먼지) 데이터

In [74]:
import pandas as pd
import numpy as np

#### 가) 데이터 불러오기 및 열 이름 변경

##### a) 다운받은 csv 형식의 파일을 pnadas로 불러오세요. (변수명 : data_raw)

In [75]:
data_raw = pd.read_csv('data/서울시 대기질 자료 제공_2020-2021.csv', encoding = 'cp949')
data_raw.head(3)

일시   구분  미세먼지(PM10)  초미세먼지(PM25)
0  2021-05-31 23:00   평균        43.0         23.0
1  2021-05-31 23:00  강남구        42.0         24.0
2  2021-05-31 23:00  강동구        47.0         30.0

##### b) 데이터셋의 속성을 확인하세요. 위 데이터는 몇행 몇열로 이루어져 있나요?

In [76]:
data_raw.shape

(322606, 4)

A. 322606행, 4열

##### c) 열 이름을 아래와 같이 변경하세요.
일시 : DateTime, 구분 : Gu, 미세먼지(PM10) : PM10, 초미세먼지(PM25) : PM2.5

In [77]:
data_raw.columns = ['DateTime', 'Gu', 'PM10', 'PM2.5']
data_raw.columns

Index(['DateTime', 'Gu', 'PM10', 'PM2.5'], dtype='object')

In [78]:
data_raw.rename(columns = {'일시':'DateTime', '구분':'Gu', '미세먼지(PM10)':'PM10', '초미세먼지(PM25)':'PM2.5'}, inplace = True)
data_raw.columns

Index(['DateTime', 'Gu', 'PM10', 'PM2.5'], dtype='object')

`.columns`는 모든 column의 갯수만큼 반환해야하고, `rename`은 필요한 피쳐의 이름만 변경 가능

#### 나) 변환하기

##### a) 위에서 'DateTime' 열의 데이터 클래스는 무엇인가요? (문자 or 숫자)

In [79]:
data_raw.dtypes

DateTime     object
Gu           object
PM10        float64
PM2.5       float64
dtype: object

A. 문자 클래스

##### b) 아래 pandas의 to_datetime 함수를 이용하여 시간 타입(datetime)으로 변환하여 datetime이라는 변수를 추가하세요.

In [80]:
data_raw['datetime'] = pd.to_datetime(data_raw['DateTime'], format = '%Y-%m-%d %H:%M')
data_raw['datetime']

0        2021-05-31 23:00:00
1        2021-05-31 23:00:00
2        2021-05-31 23:00:00
3        2021-05-31 23:00:00
4        2021-05-31 23:00:00
                 ...        
322601   2020-01-01 00:00:00
322602   2020-01-01 00:00:00
322603   2020-01-01 00:00:00
322604   2020-01-01 00:00:00
322605   2020-01-01 00:00:00
Name: datetime, Length: 322606, dtype: datetime64[ns]

##### c) 변경 후, "datetime" 열의 데이터 클래스를 다시 확인하세요.

In [81]:
data_raw.dtypes

DateTime            object
Gu                  object
PM10               float64
PM2.5              float64
datetime    datetime64[ns]
dtype: object

A. datetime64[ns] 클래스

##### d) "datetime" 열에서 연월일(yyyymmdd)와 시분(hhmm)을 추출하여 새로운 열을 아래와 같이 추가하세요.

`strftime()` 메소드는 Datetime을 형식 코드로 입력, 출력에 지정된 특정 형식을 나타내는 문자열을 반환

In [82]:
data_raw['yyyymmdd'] = data_raw['datetime'].dt.strftime('%Y%m%d')
data_raw['hhmm'] = data_raw['datetime'].dt.strftime('%H%M')
data_raw.head(3)

DateTime   Gu  PM10  PM2.5            datetime  yyyymmdd  hhmm
0  2021-05-31 23:00   평균  43.0   23.0 2021-05-31 23:00:00  20210531  2300
1  2021-05-31 23:00  강남구  42.0   24.0 2021-05-31 23:00:00  20210531  2300
2  2021-05-31 23:00  강동구  47.0   30.0 2021-05-31 23:00:00  20210531  2300

##### e) 'datetime' 열은 drop 함수를 사용하여 삭제하여 줍니다.

In [83]:
data_raw = data_raw.drop(['datetime'], axis = 1)

#### 다) 분석하기

##### a) "data_raw"에서 추출한 데이터에서 서울시 시간별 모든 지역(미세먼지, 초미세먼지)의 평균값만을 추출하세요.

In [84]:
data_raw_ave = data_raw[data_raw["Gu"] == "평균"]
data_raw_ave.head(3)

DateTime  Gu  PM10  PM2.5  yyyymmdd  hhmm
0   2021-05-31 23:00  평균  43.0   23.0  20210531  2300
26  2021-05-31 22:00  평균  43.0   23.0  20210531  2200
52  2021-05-31 21:00  평균  45.0   25.0  20210531  2100

##### b) a)에서 추출한 데이터를 시간에 따라 오름차순으로 정렬하세요.

In [85]:
data_raw_ave = data_raw_ave.sort_values("DateTime")
data_raw_ave.head(3)

DateTime  Gu  PM10  PM2.5  yyyymmdd  hhmm
322580   2020-01-01 0:00  평균  22.0   15.0  20200101  0000
322320  2020-01-01 10:00  평균  28.0   18.0  20200101  1000
322294  2020-01-01 11:00  평균  31.0   20.0  20200101  1100

##### c) 정렬 후 인덱스가 섞이기 때문에, 인덱스를 초기화 해주세요.

In [86]:
data_raw_ave = data_raw_ave.reset_index(drop = True)
data_raw_ave.head(3)

DateTime  Gu  PM10  PM2.5  yyyymmdd  hhmm
0   2020-01-01 0:00  평균  22.0   15.0  20200101  0000
1  2020-01-01 10:00  평균  28.0   18.0  20200101  1000
2  2020-01-01 11:00  평균  31.0   20.0  20200101  1100

##### d) 모든 날짜에 대한 일평균 PM10, PM2.5를 계산하세요.

In [87]:
data_raw_ave_group = data_raw_ave.groupby('yyyymmdd', as_index = True).mean()
data_raw_ave_group.head(3)

PM10      PM2.5
yyyymmdd                      
20200101  32.916667  22.458333
20200102  58.375000  39.958333
20200103  67.166667  45.583333

##### e) 환경부는 아래와 같은 기준으로 미세먼지 농도별 예보등급을 구분하였다. 아래 기준 (VB : Very Bad, B : Bad, N : Normal, G : Good)으로 구분하여 새로운 열을 추가하세요.

매우나쁨(Very Bad) : PM10 151㎍/㎥ 이상

나쁨(Bad) : PM10 81~150㎍/㎥

보통(Normal) : PM10 31~80㎍/㎥

좋음(Good) : PM10 0~30㎍/㎥

In [88]:
pm = [(data_raw_ave_group['PM10'] >= 151),
    (data_raw_ave_group['PM10'] >= 81) & (data_raw_ave_group['PM10'] <= 150),
    (data_raw_ave_group['PM10'] >= 31) & (data_raw_ave_group['PM10'] <= 80),
    (data_raw_ave_group['PM10'] >= 0) & (data_raw_ave_group['PM10'] <= 30)]

val = ['VB', 'B', 'N', 'G']

data_raw_ave_group['등급'] = np.select(pm, val)

In [89]:
data_raw_ave_group.head(3)

PM10      PM2.5 등급
yyyymmdd                         
20200101  32.916667  22.458333  N
20200102  58.375000  39.958333  N
20200103  67.166667  45.583333  N

##### f) 세계보건기구는 2013년 발암물질로 규정된 미세먼지와 초미세먼지의 위험성에 주목해 두 물질의 기준을 강화했다. 미세먼지 기준은 연간 평균 15㎍/㎥ 이하, 24시간 평균 45㎍/㎥ 이하로 강화했다. 일평균 미세먼지(PM10)가 45㎍/㎥ 이상인 날은 몇일인가요?

In [90]:
(data_raw_ave_group['PM10'] >= 45).sum()

154

A. 154일

#### 라) 월별 파일 분류하기 및 분석

`data_raw_ave`를 월별로 분류하여 csv형식으로 저장하세요.

a) index를 연월(%Y%mm%d)로 변경하여 `datetime` 열을 추가

In [91]:
data_raw_ave_group['datetime'] = pd.to_datetime(data_raw_ave_group.index, format = '%Y%m%d')
data_raw_ave_group.head(3)

PM10      PM2.5 등급   datetime
yyyymmdd                                    
20200101  32.916667  22.458333  N 2020-01-01
20200102  58.375000  39.958333  N 2020-01-02
20200103  67.166667  45.583333  N 2020-01-03

##### b-1) datetime 열에서 년월만 추출하여 'yyyymm' 열을 추가

In [92]:
data_raw_ave_group['yyyymm'] = data_raw_ave_group['datetime'].dt.strftime('%Y%m')
data_raw_ave_group.head(3)

PM10      PM2.5 등급   datetime  yyyymm
yyyymmdd                                            
20200101  32.916667  22.458333  N 2020-01-01  202001
20200102  58.375000  39.958333  N 2020-01-02  202001
20200103  67.166667  45.583333  N 2020-01-03  202001

##### b-2) datetime 열에서 일만 추출하여 'day' 열을 추가

In [93]:
data_raw_ave_group['day'] = data_raw_ave_group['datetime'].dt.strftime('%d')
data_raw_ave_group.head(3)

PM10      PM2.5 등급   datetime  yyyymm day
yyyymmdd                                                
20200101  32.916667  22.458333  N 2020-01-01  202001  01
20200102  58.375000  39.958333  N 2020-01-02  202001  02
20200103  67.166667  45.583333  N 2020-01-03  202001  03

#### c) 사용한 필요없는 datetime 열 제거

In [94]:
data_raw_ave_group = data_raw_ave_group.drop(['datetime'], axis = 1)
data_raw_ave_group.head(3)

PM10      PM2.5 등급  yyyymm day
yyyymmdd                                     
20200101  32.916667  22.458333  N  202001  01
20200102  58.375000  39.958333  N  202001  02
20200103  67.166667  45.583333  N  202001  03

##### d) 월별로 데이터 저장

In [95]:
data_uni = data_raw_ave_group['yyyymm'].unique()
data_uni

array(['202001', '202002', '202003', '202004', '202005', '202006',
       '202007', '202008', '202009', '202010', '202011', '202012',
       '202101', '202102', '202103', '202104', '202105'], dtype=object)

In [ ]:
for i in data_uni :
    data_month = data_raw_ave_group.loc[data_raw_ave_group['yyyymm'] == i, :]
    data_month.to_csv(f'PM_{i}.csv', index = False)

In [105]:
for i in data_raw_ave_group.groupby('yyyymm') :
    i[1].to_csv(f'PM_{i[0]}.csv', index = False)